### PR TITLE
[Qt] payment request / server work - part 2

### PR DESCRIPTION
**Pull request by Diapolo**
This continues my work on the payment request / server part of our BIP70 implementation.
Please see individual commits for detailed descriptions.

### CODE REVIEW

**Comment by Diapolo** [comment]
I'd love to see that reviewed or ACKd @laanwj @gavinandresen, mind taking a look?

File: src/qt/paymentrequestplus.h [hunk 1]
```diff
@@ -27,7 +27,6 @@ class PaymentRequestPlus
     bool SerializeToString(std::string* output) const;
 
     bool IsInitialized() const;
-    QString getPKIType() const;
```
Comment (jonasschnelli) [inline src/qt/paymentrequestplus.h]: This method is unused. Is there a plan to make use of it soon? IMO we should keep unused implementation in master at minimum. Better keep this in a separate branch and add it if we use it.

**Comment by jonasschnelli** [comment]
Needs rebase.

File: src/qt/paymentrequestplus.h (hunk 1 above)
Comment (jonasschnelli) [inline src/qt/paymentrequestplus.h]: Nevermind.
You removed it. Sorry.

**Comment by jonasschnelli** [comment]
It's not related to this pull, but better here than in a new issue:
I did try creating a payment-request over gavins php tool and selected bitcoincore as merchant.
I got a `GUI: PaymentRequestPlus::getMerchant : Payment request: empty certificate chain`.
A) is there no merchant cert behind "bitcoincore"?
B) IMO we should show something to the user in case of a invalid or missing cert chain.

**Comment by Diapolo** [comment]
@jonasschnelli Rebased

As for the cert problems with gavins php tool, IMHO you won't get any authenticated payment request from it and consensus then is to silently treat it as valid unauthenticated request. I had that discussion with @laanwj @gavinandresen and @mikehearn ;).

**Comment by jonasschnelli** [comment]
Tested, reviewed, stepped.
Mostly move-only, cleanup things (does that require 7 commits, squash?).

ACK

**Comment by Diapolo** [comment]
I did each change in a seperate commit to make it easy to review :).

**Comment by luke-jr** [comment]
Does anyone have a test plan for this?

**Comment by Diapolo** [comment]
@luke-jr Are you serious??

File: src/qt/paymentrequestplus.cpp [hunk 2]
```diff
@@ -125,7 +119,7 @@ bool PaymentRequestPlus::getMerchant(X509_STORE* certStore, QString& merchant) c
     // The first cert is the signing cert, the rest are untrusted certs that chain
     // to a valid root authority. OpenSSL needs them separately.
     STACK_OF(X509) *chain = sk_X509_new_null();
-    for (int i = certs.size()-1; i > 0; i--) {
+    for (size_t i = certs.size() - 1; i > 0; i--) {
```
Comment (laanwj) [inline src/qt/paymentrequestplus.cpp]: NACK on changing this. There is check for certs.empty() that prevents this from becoming a bug, but for loops that count down, I generally prefer to use signed types. In this case, `i` will never visit 0, but normally a loop like this would look like:

    for (int i = certs.size() - 1; i >= 0; i--) {

If you then change the int to e.g. size_t, it will loop forever.

Comment (Diapolo) [inline src/qt/paymentrequestplus.cpp]: Alright, will revert this change here.

**Comment by Diapolo** [comment]
@jonasschnelli If you've got an idea how to group the commits so they can get merged I'll gladly do this. Guess this needs an ACK from @laanwj still :).

**Comment by Diapolo** [comment]
@laanwj @jonasschnelli Is there some consensus possible to at least pick most of these commits or some, so we can get on here!?

**Comment by Diapolo** [comment]
I'd like to improve payment request stuff, but not getting on with this is blocking you know ;).

**Comment by laanwj** [comment]
ACK apart from the char* change.

**Comment by Diapolo** [comment]
@laanwj Thanks for review, I fixed the spelling error and remove the QVariant commit, should be ready now :).
